ttnn/cpp/ttnn/operations/data_movement/transpose/device/transpose_program_factory.cpp

```diff
@@ -1911,11 +1931,22 @@ operation::ProgramWithCallbacks transpose_wh_multi_core_sharded(const Tensor& a,
                 program, cb_output, *dst_buffer, num_tiles_per_shard * dst_single_tile_size);
         }
 
-        uint32_t Wt = shard_spec.shape[1] / TILE_WIDTH;
-        uint32_t Ht = src_tensor.get_legacy_shape()[-2] / TILE_HEIGHT;
-        uint32_t HtWt = Ht * Wt;
-        uint32_t N = shard_spec.shape[0] / src_tensor.get_legacy_shape()[-2];
-        uint32_t NHtWt = N * HtWt;
+        auto padded_shape = src_tensor.get_padded_shape();
+        auto shard_shape = shard_spec.shape;
+
+        uint32_t H = padded_shape[2], W = padded_shape[3];
+        uint32_t Hs = shard_shape[0], Ws = shard_shape[1];
+
+        uint32_t Hts = Hs / tile.tile_shape[0];
+        uint32_t Wts = Ws / tile.tile_shape[1];
+
+        uint32_t Ht = H / tile.tile_shape[0];
+        uint32_t Ht_per_shard = std::min(Ht, Hts);
+
+        uint32_t num_hw_blocks_per_shard = Hts > Ht ? Hts / Ht : 1;
+
+        uint32_t HtWt_tile_size = Ht_per_shard * Wts;
+        uint32_t num_blocks = num_hw_blocks_per_shard * HtWt_tile_size;
```

Comment:
I also see similar code often enough. Wonder if something about it must be easy to query from the spec.

Reply:
Some of these like Hts/Wts would definitely be good to have. Num shards along height and num shards along width are calculated everywhere.